cmd/katapult/vms_test.go

```diff
@@ -0,0 +1,583 @@
+) {
+	switch {
+	case ref.ID != "":
+		delete(v.powerStates, "i"+ref.ID)
+	case ref.FQDN != "":
+		delete(v.powerStates, "s"+ref.FQDN)
+	}
+	return nil, nil, nil
+}
+
+func TestVMs_List(t *testing.T) {
+	tests := []struct {
+		name string
+
+		id         map[string]vmPages
+		subdomains map[string]vmPages
+
+		args    []string
+		want    string
+		stderr  string
+		wantErr string
+	}{
+		{
+			name: "test un-paginated vm ID list",
+			id: map[string]vmPages{
+				"1": {
+					{
+						{
+							ID:          "1",
+							Name:        "test",
+							Hostname:    "test.example.com",
+							FQDN:        "test.example.com",
+							Description: "test",
+							Package:     &core.VirtualMachinePackage{Name: "test"},
+						},
+					},
+				},
+			},
+			args: []string{"list", "--id=1"},
+			want: " - test (test.example.com) [1]: test\n",
+		},
+		{
+			name: "test un-paginated vm subdomain list",
+			subdomains: map[string]vmPages{
+				"1": {
+					{
+						{
+							ID:          "1",
+							Name:        "test",
+							Hostname:    "test.example.com",
+							FQDN:        "test.example.com",
+							Description: "test",
+							Package:     &core.VirtualMachinePackage{Name: "test"},
+						},
+					},
+				},
+			},
+			args: []string{"list", "--subdomain=1"},
+			want: " - test (test.example.com) [1]: test\n",
+		},
+		{
+			name: "test paginated vm list",
+			subdomains: map[string]vmPages{
+				"1": {
+					{
+						{
+							ID:          "0",
+							Name:        "test",
+							Hostname:    "test.example.com",
+							FQDN:        "test.example.com",
+							Description: "test",
+							Package:     &core.VirtualMachinePackage{Name: "test"},
+						},
+					},
+					{
+						{
+							ID:          "1",
+							Name:        "test1",
+							Hostname:    "test1.example.com",
+							FQDN:        "test1.example.com",
+							Description: "test1",
+							Package:     &core.VirtualMachinePackage{Name: "test1"},
+						},
+					},
+					{
+						{
+							ID:          "2",
+							Name:        "test2",
+							Hostname:    "test2.example.com",
+							FQDN:        "test2.example.com",
+							Description: "test2",
+							Package:     &core.VirtualMachinePackage{Name: "test2"},
+						},
+					},
+				},
+			},
+			args: []string{"list", "--subdomain=1"},
+			want: ` - test (test.example.com) [0]: test
+ - test1 (test1.example.com) [1]: test1
+ - test2 (test2.example.com) [2]: test2
+`,
+		},
+		{
+			name:    "test not found",
+			args:    []string{"list", "--subdomain=not_exists"},
+			wantErr: "katapult: not_found",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			cmd := virtualMachinesCmd(&vmsClient{organizationIDPages: tt.id, organizationSubdomainPages: tt.subdomains})
+			cmd.SetArgs(tt.args)
+			assertCobraCommand(t, cmd, tt.wantErr, tt.want, tt.stderr)
+		})
+	}
+}
+
+func TestVMs_Poweroff(t *testing.T) {
+	successMessage := "Virtual machine successfully powered down.\n"
+	tests := []struct {
```

Comment:
There's no test case for when no VM ID or FQDN is provided. The other power-state tests are also missing such a test case.